ipld/merkledag/merkledag.go

```diff
@@ -19,16 +18,14 @@ import (
 	basicnode "github.com/ipld/go-ipld-prime/node/basic"
 )
 
-// TODO: We should move these registrations elsewhere. Really, most of the IPLD
-// functionality should go in a `go-ipld` repo but that will take a lot of work
-// and design.
-func init() {
-	format.Register(cid.DagProtobuf, DecodeProtobufBlock)
-	format.Register(cid.Raw, DecodeRawBlock)
-	format.Register(cid.DagCBOR, ipldcbor.DecodeBlock)
+var ipldLegacyDecoder *legacy.Decoder
 
-	legacy.RegisterCodec(cid.DagProtobuf, dagpb.Type.PBNode, ProtoNodeConverter)
-	legacy.RegisterCodec(cid.Raw, basicnode.Prototype.Bytes, RawNodeConverter)
+// TODO: Don't require global registries
+func init() {
+	d := legacy.NewDecoder()
+	d.RegisterCodec(cid.DagProtobuf, dagpb.Type.PBNode, ProtoNodeConverter)
+	d.RegisterCodec(cid.Raw, basicnode.Prototype.Bytes, RawNodeConverter)
+	ipldLegacyDecoder = d
```

Comment:
@willscott thanks for doing https://github.com/ipfs/go-merkledag/pull/104. I mostly copied it. Although I left the globals up in the global space as a reminder for us to try and exit globals land when it becomes possible 😅

Reply:
> as a reminder for us to try and exit globals land when it becomes possible 

I'm confused by this - in 104 i did exit globals land. - is there a reason not to stay consistent with that in the copy here, since it's the direction we want to go anyway?

Reply:
> I'm confused by this - in 104 i did exit globals land

We're still stuck with globals:
1. The merkledag libraries still have globals. IMO this is bad for the same reasons the other ones were bad, it just turns out the other ones are basically unused independently so for now we can coalesce here
2. There are still go-ipld-prime globals used implicitly in go-ipld-legacy unless we fix that https://github.com/ipfs/go-ipld-legacy/pull/12/files#r1211104340
    - Ran out of time to do the kubo PR exercising this (https://github.com/ipfs/kubo/pull/9907). Maybe it'll be fine to leave this merkledag PR mostly alone (aside from passing in something like a linksystem to go-ipld-legacy). However, to start wrestling control away from the globals we might also want to make the legacy decoder used by merkledag publicly settable (albeit with the various warnings about being careful about using it, and trying to only use it in top-level binaries).
    
>  is there a reason not to stay consistent with that in the copy here

If we go with making the go-ipld-legacy decoder publicly settable (and therefore not tied to go-ipld-prime's global registry) then this approach looks much better rather than just being a stylistic difference. If it's just a stylistic difference then 🤷.